---
title: "Homework4_JKGWAS"
author: "Pabitra Joshi and Lindsey Kornowske"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  pdf_document:
    keep_tex: true
    toc: yes
    toc_depth: '6'
  word_document:
    toc: yes
    toc_depth: '6'
---

\begin{figure*}[htbp]
\begin{center}
\includegraphics[width = 0.5\textwidth]{JKGWAS_logo.png}
\end{center}
\end{figure*}

# Data and Functions

The class maize datasets were used for this assignment. They are available from "http://www.zzlab.net" 

```{r}
source("http://www.zzlab.net/StaGen/2021/R/G2P.R")
source("http://www.zzlab.net/StaGen/2021/R/GWASbyCor.R")
library(JKGWAS)

# Genotype Data
X = read.csv(file = "./../datasets/mdp_numeric.txt", header = TRUE, sep ="");
# Phenotype Data
y = read.csv(file = "./../datasets/CROPS545_Phenotype.txt", header = TRUE, sep = "");
# Covariates Data
CV = read.csv(file = "./../datasets/CROPS545_Covariates.txt", header = TRUE, sep = "");
# SNP information data
SNP = read.csv(file = "./../datasets/mdp_SNP_information.txt", header = TRUE, sep = "");

```

# Question 1 and 2
\textbf{(1)	The package should contain at least three input: y, X , and C that are R objects of numeric data frame. Their dimensions are n by 1, n by m, and n by t corresponding to phenotype, genotype and covariate data, where n is number of individuals, m is number of markers, and t is number of covariates. The function should return probability values with dimension of 1 by m for the association tests between phenotype and markers. Markers are tested one at a time with covariates in C included as covariates (15 points).
(2)	The package should perform PCA and incorporate PCs as cofactors for GWAS.  Your package should also automatically exclude the PCs that are in linear dependent to the covariates provided by users. (25 points).}

## JKGWAS Summary

The JKGWAS Package contains four functions that are summarized briefly as follows, more information is located in the JKGWAS Package documentation: \par
\textbullet JKPCA takes genotype (X) data and covariate data (CV), computes the PCA on X, then automatically removes PCs that are linearly dependent to the CVs by method of comparing matrix rank. PCs are removed from the matrix in succesion and those that do not change the rank by removal are determined to be linearly independent because they do not provide additional information. \par
\textbullet JKGLM takes phenotype (y), genotype (X), covariate (CV), and principal component (PC) inputs (ideally provided from JKPCA) and returns p-values calculated for the association tests between the phenotype and SNPs \par
\textbullet JKQQ takes the pvalues from JKGLM and visualizes them by QQ plot. Expected p-values of length m are simulated from the continuous distribution.\par
\textbullet JKManhattan visualizes the pvalues from JKGLM by Manhattan plot. User input QTNs can also be visualized. The significance threshold can be set, or it will default to Bonferoni correction for alpha = 0.05 \par



# Question 3

\textbf{(3)	Develop a user manual and tutorials. Name your package and create a logo. (20 points).}

The JKGWAS package is named for Pabitra Joshi and Lindsey Kornowske, the label is displayed in Figure \ref{fig:logo}.

\begin{figure*}[htbp]
\begin{center}
\includegraphics[width = 0.3\textwidth]{JKGWAS_logo.png}
  \caption{JKGWAS Package Logo}
  \label{fig:logo}
\end{center}
\end{figure*}

The JKGWAS package documentation is provided in a separate file, "JKGWAS$\_$0.1.0.pdf" and further information about its use are provided in the user tutorial, "JKGWAS$\_$UserTutorial.pdf"

# Question 4

\textbf{(4)	Perform GWAS on the data provided or your own data which must contain cofactors (15 points).}

```{r}
## Get Principal Components with JKPCA()
PC = JKPCA(X, CV, npc = 10);

## Perform GWAS by GLM with JKGLM()
Pvals = JKGLM(X = X, y = y, CV, PC);

## Visualize GWAS by QQ Plot with JKQQ()
JKQQ(Pvals);

## Visualize GWAS by Manhattan Plot with JKManhattan()
JKManhattan(Pvals = Pvals, SNP = SNP,sigcutoff = NULL );

```
First, GWAS was performed with the phenotype data provided. In the Manhattan plot below, we can see that 3 SNP were detected, but because we do not have information about the QTNs, we do not know whether these significant observations represent true positives or not. Next, we use a simulated phenotype to better assess the performance of the JKGWAS approach.

## QTN Simulation

```{r}
G2P.sim = G2P(X= X,
             h2= 0.75,
             alpha=1,
             NQTN=qtn,
             distribution="norm");

G2P.y = as.data.frame(G2P.sim$y);

Pvals.sim = JKGLM(X = X, y = G2P.y, CV, PC);

## Visualize GWAS by QQ Plot with JKQQ()
JKQQ(Pvals.sim);

## Visualize GWAS by Manhattan Plot with JKManhattan()
JKManhattan(Pvals = Pvals.sim, SNP = SNP, sigcutoff = NULL, QTN = G2P.sim$QTN.position );

```

If we use a simulated phenotype instead of the provided phenotype data, 50% of our 10 QTN are detected correctly. With 5 QTN detected out of 8 significant SNPs total, that is a True Positive rate of 62.5%.

# Question 5

\textbf{(5)	Demonstrate that your method is superior to the competing method (GWASbyCor) through simulation with at least 30 replicates (25 points).}

In order to compare GWASbyCor and GWASbyGLM, we created a function called compareGWASnTimes. The function arguments are: \par
\texbullet n, the number of times to run the simulation \par
\textbullet X, the numeric genomic data \par
\textbullet qtn, the number of qtns to be simulated \par
\textbullet CV, the covariate matrix to be passed to the JKGLM function. The default value is NULL \par
\textbullet PC, the principal component matrix to be passed to the JKGLM function. The default value is NULL \par

For each iteration, the G2P function simulates the phenotype for X with a heritability of 0.75. Then, the output phenotype is used to compute the GWAS by cor and the GWAS by GLM. The number of True Positives, as well as the True Positive Rate, which is calculated as the number of QTNs that is correctly identified out of all significant SNPs (p-value is smaller than 0.05/total pvalues, Bonferroni correction is automatic). These two dataframes are output as a list, where the first item is the count of true positive QTNs and the second item is the true positive rate.

```{r}
compareGWASnTimes = function(n = 10, X, qtn = 10, CV = NULL, PC = NULL) {
   X = dplyr::select_if(X, is.numeric);
   # create array to store average for each iteration
  store.truepos = data.frame();
  store.rate = data.frame();

  for(i in 1:n){
# first, simulate phenotype
   # same parameters as Q2-4
G2P.sim = G2P(X= X,
             h2= 0.75,
             alpha=1,
             NQTN=qtn,
             distribution="norm");

# get QTN positions in GD data
G2P.sim.qtn = G2P.sim$QTN.position;

#get vector of pvals
#by correlation
p.list.cor = GWASbyCor(X, G2P.sim$y);
p.list.cor = as.matrix(t(p.list.cor));

y = as.data.frame(G2P.sim$y);
#by glm
p.list.glm = JKGLM(X, y, CV, PC);
p.list.glm = as.matrix(p.list.glm)

p.df = as.data.frame(cbind(p.list.cor, p.list.glm));
names(p.df) = c("COR", "GLM");

nc = ncol(p.df);
for(j in 1:nc){
  p.index = p.df[,j];
#identify qtn pvals
qtn.p = p.index[G2P.sim.qtn];
  #sort vector of pvals as increasing
p.index.sort =sort(p.index, decreasing = FALSE);
p.sig.cor = p.index.sort[p.index.sort < 0.05/length(p.index.sort)];


#find all qtn pvals in total list
truePos.cor=intersect(p.sig.cor, qtn.p);
#find all significant pvals not in qtn list
falsePos.cor=setdiff(p.sig.cor, qtn.p);
#save the number of true positives for this iteration
store.truepos[i,j] = length(truePos.cor);
store.rate[i,j] = length(truePos.cor)/length(p.sig.cor);

  }

}
names(store.truepos) = c("COR", "GLM");
names(store.rate) = c("COR", "GLM");

return(list(store.truepos, store.rate));

#summary(store.truepos);
}

# Run the simulation comparison 30 times
set.seed(1348);
test = compareGWASnTimes(n=30, X = X, qtn = 10);
```

Now that we have stored our true positive rates for each method in a variable "test", we can compute the summary statistics.

```{r}
mean(test[[2]]$COR);
sd(test[[2]]$COR);
mean(test[[2]]$GLM);
sd(test[[2]]$GLM);
```

By these statistics, we can see that the standard deviation of the true positive rate for each method is comparable. The True Positive Rate is higher for the GWAS by GLM than the GWAS by Cor.


# Extra Credit

\textbf{(6)	Demonstrate that your package is better than BLINK C version (http://zzlab.net/blink) on either statistical power or speed (25 points). }

```{r}
```

